src/libraries/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.ParseMapModules.cs

```diff
@@ -0,0 +1,135 @@
+            if (pos > 0)
+            {
+                if (long.TryParse(s.AsSpan(start, pos), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out long startingAddress) &&
+                    long.TryParse(s.AsSpan(pos + 1, end - (pos + 1)), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out long endingAddress))
+                {
+                    return (startingAddress, (int)(endingAddress - startingAddress));
+                }
+            }
+
+            return default;
+        }
+
+        private static bool HasReadAndExecFlags(string s, ref int start, ref int end)
+        {
+            bool sawRead = false, sawExec = false;
+            for (int i = start; i < end; i++)
+            {
+                if (s[i] == 'r')
+                    sawRead = true;
+                else if (s[i] == 'x')
+                    sawExec = true;
+            }
+
+            return sawRead & sawExec;
+        }
+
+        internal static ProcessModuleCollection? ParseMapsModules(int pid)
+        {
+            try
+            {
+                return ParseMapsModulesCore(File.ReadLines(GetMapsFilePathForProcess(pid)));
+            }
+            catch (IOException) { }
+            catch (UnauthorizedAccessException) { }
+
+            return null;
+        }
+
+        private static ProcessModuleCollection ParseMapsModulesCore(IEnumerable<string> lines)
+        {
+            Debug.Assert(lines != null);
+
+            ProcessModule? previous = null;
+            ProcessModuleCollection modules = new(0);
+
+            // Parse each line from the maps file into a ProcessModuleCollection result
+            foreach (string line in lines)
+            {
+                // Use a StringParser to avoid string.Split costs
+                var parser = new StringParser(line, separator: ' ', skipEmpty: true);
+
+                // Parse the address start and size
+                (long Start, int Size) addressRange = parser.ParseRaw(TryParseAddressRange);
+
+                if (addressRange == default)
+                {
+                    continue;
+                }
+
+                // Parse the permissions
+                bool hasReadAndExecFlags = parser.ParseRaw(HasReadAndExecFlags);
+
+                // Skip past the offset, dev, and inode fields
+                parser.MoveNext();
+                parser.MoveNext();
+                parser.MoveNext();
+
+                // Parse the pathname
+                if (!parser.MoveNext())
+                {
+                    continue;
+                }
+
+                string pathname = parser.ExtractCurrentToEnd();
+                bool isContinuation = previous?.FileName == pathname && (long)previous.BaseAddress + previous.ModuleMemorySize == addressRange.Start;
+
+                // we only care about entries with 'r' and 'x' set, unless it is a continuation of previous.
+                if (!hasReadAndExecFlags && !isContinuation)
```

Comment:
Isn't this still skipping the first non r-x sections?

Reply:
Yes, but that's ok? The assumption is that first r+x section is start of the range.

Reply:
I have moved `isContinuation` up so we don't need to check it again.

Reply:
> The assumption is that first r+x section is start of the range.

That's a false assumption.

```
7ffff72ff000-7ffff7371000 r--p 00000000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
7ffff7371000-7ffff7657000 r-xp 00072000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
7ffff7657000-7ffff7659000 rwxp 00358000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
7ffff7659000-7ffff780b000 r-xp 0035a000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
7ffff780b000-7ffff7812000 r--p 0050c000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
7ffff7812000-7ffff7813000 r--p 00513000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
7ffff7813000-7ffff7958000 r--p 00514000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
7ffff7958000-7ffff797c000 r--p 00658000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
7ffff797c000-7ffff7994000 rw-p 0067c000 fd:00 18097725                   /usr/lib64/dotnet/shared/Microsoft.NETCore.App/5.0.0/libcoreclr.so
```

Reply:
Given:
```
A-B       r---
B-C       r-x-
C-D       x-r-
D-E       rw--
F-G       r--x
H-I       ---r
J-K       -r--
```
* Today we are accounting for B to D
  * The assumption "where to start" exists, it is unchanged.
* PR is accounting for B to K
  * This fix is focused on "do not skip adjacent ranges (once started)".

I guess your idea is that we should simply remove the r+x filter and account for A to K. Ok, this is not what I thought in #45180. I am not clear why we had r+x filter in the first place?

Reply:
> PR is accounting for B to K 

https://github.com/dotnet/runtime/issues/45482 is about A not getting accounted. Let's fix that here too instead of making another PR touching/reviewing the same code.

> I am not clear why we had r+x filter in the first place?

To not include regular files that are mapped into the process in `Process.Modules`.

Reply:
Fixed in 9fdc4eb.